src/core/ext/transport/chttp2/transport/frame.cc

```diff
@@ -0,0 +1,519 @@
+                                std::move(payload)};
+}
+
+absl::StatusOr<Http2RstStreamFrame> ParseRstStreamFrame(
+    const Http2FrameHeader& hdr, SliceBuffer& payload) {
+  if (hdr.flags != 0) {
+    return absl::InternalError(
+        absl::StrCat("unsupported rst stream flags: ", hdr.ToString()));
+  }
+
+  if (payload.Length() != 4) {
+    return absl::InternalError(
+        absl::StrCat("invalid rst stream payload: ", hdr.ToString()));
+  }
+
+  if (hdr.stream_id == 0) {
+    return absl::InternalError(
+        absl::StrCat("invalid stream id: ", hdr.ToString()));
+  }
+
+  uint8_t buffer[4];
+  payload.CopyToBuffer(buffer);
+
+  return Http2RstStreamFrame{hdr.stream_id, Read4b(buffer)};
+}
+
+absl::StatusOr<Http2SettingsFrame> ParseSettingsFrame(
+    const Http2FrameHeader& hdr, SliceBuffer& payload) {
+  if (hdr.stream_id != 0) {
+    return absl::InternalError(
+        absl::StrCat("invalid stream id: ", hdr.ToString()));
+  }
+  if (hdr.flags == kFlagAck) return Http2SettingsFrame{true, {}};
+  if (hdr.flags != 0) {
+    return absl::InternalError(
+        absl::StrCat("invalid settings flags: ", hdr.ToString()));
+  }
+
+  if (payload.Length() % 6 != 0) {
+    return absl::InternalError(
+        absl::StrCat("invalid settings payload: ", hdr.ToString(),
+                     " -- settings must be multiples of 6 bytes long"));
+  }
+
+  Http2SettingsFrame frame{false, {}};
+  while (payload.Length() != 0) {
+    uint8_t buffer[6];
+    payload.MoveFirstNBytesIntoBuffer(6, buffer);
+    frame.settings.push_back({
+        Read2b(buffer),
+        Read4b(buffer + 2),
+    });
+  }
+  return std::move(frame);
+}
+
+absl::StatusOr<Http2PingFrame> ParsePingFrame(const Http2FrameHeader& hdr,
+                                              SliceBuffer& payload) {
+  if (payload.Length() != 8) {
+    return absl::InternalError(
+        absl::StrCat("invalid ping payload: ", hdr.ToString()));
+  }
+
+  if (hdr.stream_id != 0) {
+    return absl::InternalError(
+        absl::StrCat("invalid ping stream id: ", hdr.ToString()));
+  }
+
+  bool ack;
+  switch (hdr.flags) {
+    case 0:
+      ack = false;
+      break;
+    case kFlagAck:
+      ack = true;
+      break;
+    default:
+      return absl::InternalError(
+          absl::StrCat("invalid ping flags: ", hdr.ToString()));
+  }
+
+  uint8_t buffer[8];
+  payload.CopyToBuffer(buffer);
+
+  return Http2PingFrame{ack, Read8b(buffer)};
+}
+
+absl::StatusOr<Http2GoawayFrame> ParseGoawayFrame(const Http2FrameHeader& hdr,
+                                                  SliceBuffer& payload) {
+  if (payload.Length() < 8) {
+    return absl::InternalError(
+        absl::StrCat("invalid goaway payload: ", hdr.ToString(),
+                     " -- must be at least 8 bytes"));
+  }
+
+  if (hdr.stream_id != 0) {
+    return absl::InternalError(
+        absl::StrCat("invalid goaway stream id: ", hdr.ToString()));
+  }
+
+  if (hdr.flags != 0) {
```

Comment:
As per RFC 9113, Unused flags MUST be ignored on receipt

Reply:
fixed